# 3A: Perform

In this assignment, you will demonstrate your mastery with wrangling data through different formats, with missing data, and working with text.

You can either print answers directly from your code or write them in the markdown cells below your code. Either way, make sure that your answers are visible and can be easily read in the final notebook you turn in.

## Part 1: Text wrangling and regular expressions
In this part we will work with the citation file exported from the [Nature Review Article](https://www.nature.com/articles/s41586-020-2649-2) *Array Programming with NumPy*. Below we read the file into the Python variable `cite` and print the result for you to preview.

In [2]:
# Run but do not modify this code
with open("numpy_nature.txt") as f:
    cite = f.read()

print(cite)

TY  - JOUR
AU  - Harris, Charles R.
AU  - Millman, K. Jarrod
AU  - van der Walt, Stéfan J.
AU  - Gommers, Ralf
AU  - Virtanen, Pauli
AU  - Cournapeau, David
AU  - Wieser, Eric
AU  - Taylor, Julian
AU  - Berg, Sebastian
AU  - Smith, Nathaniel J.
AU  - Kern, Robert
AU  - Picus, Matti
AU  - Hoyer, Stephan
AU  - van Kerkwijk, Marten H.
AU  - Brett, Matthew
AU  - Haldane, Allan
AU  - del Río, Jaime Fernández
AU  - Wiebe, Mark
AU  - Peterson, Pearu
AU  - Gérard-Marchant, Pierre
AU  - Sheppard, Kevin
AU  - Reddy, Tyler
AU  - Weckesser, Warren
AU  - Abbasi, Hameer
AU  - Gohlke, Christoph
AU  - Oliphant, Travis E.
PY  - 2020
DA  - 2020/09/01
TI  - Array programming with NumPy
JO  - Nature
SP  - 357
EP  - 362
VL  - 585
IS  - 7825
AB  - Array programming provides a powerful, compact and expressive syntax for accessing, manipulating and operating on data in vectors, matrices and higher-dimensional arrays. NumPy is the primary array programming library for the Python language. It has an essenti

### Question 1 (4 points)
There are several authors, each recorded on a separate line beginning with `AU`. Create a Python list of all of the author names formatted as in the file but without the extra characters and whitespace (i.e., without the `AU  - ` or the newline `\n` characters). Your list should be of the form `['Harris, Charles R.', 'Millman, K. Jarrod', ..., 'Oliphant, Travis E.']`. When you are finished, print the resulting list. 

In [3]:
import re
# Put your code to answer the question here, and store the answer as the following variable
# Please keep author names in the original order
query = re.compile(r"AU  - +.+")
list1 = re.findall(query,cite)
list2=[]
for element in list1:
    list2.append(element.replace("AU  - ",""))
q1 = list2 # List of Authors

print(q1)

['Harris, Charles R.', 'Millman, K. Jarrod', 'van der Walt, Stéfan J.', 'Gommers, Ralf', 'Virtanen, Pauli', 'Cournapeau, David', 'Wieser, Eric', 'Taylor, Julian', 'Berg, Sebastian', 'Smith, Nathaniel J.', 'Kern, Robert', 'Picus, Matti', 'Hoyer, Stephan', 'van Kerkwijk, Marten H.', 'Brett, Matthew', 'Haldane, Allan', 'del RÃ\xado, Jaime Fernández', 'Wiebe, Mark', 'Peterson, Pearu', 'Gérard-Marchant, Pierre', 'Sheppard, Kevin', 'Reddy, Tyler', 'Weckesser, Warren', 'Abbasi, Hameer', 'Gohlke, Christoph', 'Oliphant, Travis E.']


### Question 2 (4 points)
Create a Pandas DataFrame that contains three columns: one for first names, one for middle names, and one for last names for all of the authors. You should use descriptive column names, but you can use the default primary index (the row labels) of 0, 1, 2, etc. The table's rows should be in the same order as the original data set. Thus, the first few rows of your dataframe might look like the table pictured below. You are welcome to use the results of the prior question to asnwer this problem.

|      | first      | middle     | last         |
| ---- | ---------- | ---------- | ------------ |
| 0	   | Charles    | R.         | Harris       |
| 1	   | K.	        | Jarrod     | Millman      |
| 2	   | Stéfan     | J.         | van der Walt |
| 3    | Ralf       |            | Gommers      |
| 4	   | Pauli      |            | Virtanen     |   

Note that some authors do not have any middle names, in which case you can leave the middle name column blank. When you are finished, display the first 10 rows of the resulting dataframe by calling `pd.head(10)` (where Pandas has been imported as `pd`).

In [4]:
# Run but do not modify this code
import pandas as pd

In [5]:
import re
import numpy as np
# Put your code to answer the question here, and store the answer as the following variable
# Please keep author names in the original order 
df = pd.DataFrame({"first":name.split(" ")[-2] if len(name.split(" "))>2 else name.split(" ")[-1] ,"middle": name.split(" ")[-1] if len(name.split(" "))>2 else "" ,"last":name.split(",")[0]} for name in list2)




q2 = df # Pandas DataFrame with three columns

df.head(10)


first  middle          last
0    Charles      R.        Harris
1         K.  Jarrod       Millman
2    Stéfan      J.  van der Walt
3       Ralf               Gommers
4      Pauli              Virtanen
5      David            Cournapeau
6       Eric                Wieser
7     Julian                Taylor
8  Sebastian                  Berg
9  Nathaniel      J.         Smith

### Question 3 (12 points)
Below we extract the abstract from the citation and store it in a string variable `abstract`. Write regular expressions to answer the following questions about the abstract.

1. Print the starting index of everywhere `NumPy` appears in the abstract (i.e., the index of the `N` wherever `NumPy` occurs in the `abstract` string). This should be case sensitive.
2. Print all of the capitalized words in `abstract`, including words with extra capitalized letters like `NumPy`.
3. Print all of the words that immediately follow `NumPy`, but do not include the word `NumPy` itself. Note that in one occurrence it is hyphenated `NumPy-like`, in which case your code can return `-like` or `like` as you prefer.

In [6]:
# Run but do not modify this code
import re
abstract_query = re.compile(r"AB  - (.+)")
abstract = re.search(abstract_query, cite).group(1)
print(abstract)

Array programming provides a powerful, compact and expressive syntax for accessing, manipulating and operating on data in vectors, matrices and higher-dimensional arrays. NumPy is the primary array programming library for the Python language. It has an essential role in research analysis pipelines in fields as diverse as physics, chemistry, astronomy, geoscience, biology, psychology, materials science, engineering, finance and economics. For example, in astronomy, NumPy was an important part of the software stack used in the discovery of gravitational waves1 and in the first imaging of a black hole2. Here we review how a few fundamental array concepts lead to a simple and powerful programming paradigm for organizing, exploring and analysing scientific data. NumPy is the foundation upon which the scientific Python ecosystem is constructed. It is so pervasive that several projects, targeting audiences with specialized needs, have developed their own NumPy-like interfaces and array object

In [18]:
# Put your code to answer the question here, and store the answer as the following variable
# Please allow duplication in your list

q3_1 = [word.start(0)+1 for word in re.finditer(re.compile(r"( +NumPy[- ]+)"),abstract)]  # List of index

#
q3_2 = re.findall(re.compile(r"([A-Z]+[- ]\w+)"),abstract)# List of all the capitalized words


q3_3 = re.findall(re.compile(r"NumPy+(?:[ -])+([A-Za-z]+)"),abstract) # List of words that immediately follow Numpy


q3_2


[]

## Part 2: Cleaning up more system logs CSV
In this part we work with a piece of messy tabular data in the form of a poorly formatted csv file containing data about programs running on computer systems. It contains all of the data about system time and memory from Practice 3, but also includes new information about user ids and machine ids, and some data are missing in every column. (The user ids are made up and do not correspond to any real individuals).  

### Question 4 (12 points)
Below, we import the dataset using the Pandas `read_csv` function that creates a dataframe. Run the code; it will preview the first five rows. 

In [8]:
# Run but do not modify this code
import pandas as pd
import numpy as np
sys_df = pd.read_csv("more_monitor.csv")
sys_df.head()

System Time second   System Memory GB System Memory MB System User ID  \
0                       ?                  ?                ?  User ID: yw22   
1  System Time: 40 second   System Mem: 3 Gb           382 Mb              ?   
2                       ?   System Mem: 2 Gb           271 Mb   User ID: tp7   
3  System Time: 31 second   System Mem: 3 Gb           493 Mb              ?   
4  System Time: 37 second   System Mem: 3 Gb           411 Mb              ?   

       System Machine ID  
0     Machine ID: Carrot  
1                      ?  
2  Machine ID: Asparagus  
3   Machine ID: Eggplant  
4  Machine ID: Asparagus

There are several formatting issues with the default import. Address the following.

1. The `System User ID` and `System Machine ID` contain String data with the redundant information `User ID: ` and `Machine ID: ` in every row that has data. Remove these prefixes so that the columns only contain the user ids and machine ids themselves (for example, the first row should just have `yw22` in the `System User ID` column and `Carrot` in the `System Machine ID` column. 
2. The first three rows for `System Time second`, `System Memory GB` and `System Memory MB` contain numerical data but are currently formatted as strings with redundant prefix information repeating the column label and missing data represented as the string `?` instead of the Numpy `NaN` sentinel value. Fix this so that each value in the first three columns is either a single numerical value or `NaN` (note, you should use the actual `np.NaN` sentinal value, not just the String with the characters `N`, `a`, and `N`). For example, when you are done, the first three columns of the first row should all have `NaN` values, the second row should be `40`, `3`, and `382`, and so on. Note that the rows at index `400` and on have System Time recorded in minutes instead of seconds, be sure to convert these to seconds by mulitplying by 60.
3. Currently the System Memory is split accross two columns, one for the GB and one for the MB. For example, the total memory of the first program is 3 GB and 414 MB. Instead, represent the full system memory in the `System Memory GB` column, and get rid of the `System Memory MB` column. To do so, you need to convert the values in the MB column to GB (1 MB is 0.001 GB) and add that to the GB column, then use the [`drop` method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html). Missing values should remain missing after this transformation.

When you are finished, `sys_df` should have the above issues corrected. Run both of the cells with `sys_df.head()` and `sys_df.tail()` to show the first and last few rows of your dataframe.

In [9]:
# Put your code to answer the question here, and store sys_df = q4 for full credit

newData= sys_df.copy()

#Question 1
newData["System User ID"]= newData["System User ID"].str.replace("User ID: ","")
newData["System Machine ID"]= newData["System Machine ID"].str.replace("Machine ID: ","")

#Question 2
newData["System Time second"]= newData["System Time second"].str.replace("System Time: ","")
newData["System Time second"]= newData["System Time second"].str.replace("second","")
newData["System Time second"]= newData["System Time second"].str.replace("minute","")
newData["System Time second"]= newData["System Time second"].str.replace("?","0") #FIX this line of code
newData["System Time second"]= newData["System Time second"].astype(int)
newData["System Time second"]= newData["System Time second"].replace(0,np.nan)
newData["System Time second"][400:] = newData["System Time second"][400:]*60

newData["System Memory GB"]= newData["System Memory GB"].str.replace("System Mem: ","")
newData["System Memory GB"]= newData["System Memory GB"].str.replace("Gb","")
newData["System Memory GB"]= newData["System Memory GB"].str.replace("?","0") #FIX this line of code
newData["System Memory GB"]= newData["System Memory GB"].astype(int)
newData["System Memory GB"]= newData["System Memory GB"].replace(0,np.nan)

newData["System Memory MB"]= newData["System Memory MB"].str.replace("Mb","")
newData["System Memory MB"]= newData["System Memory MB"].str.replace("?","0") #FIX this line of code
newData["System Memory MB"]= newData["System Memory MB"].astype(int)
newData["System Memory MB"]= newData["System Memory MB"].replace(0,np.nan)


#Question 3
newData["System Memory GB"]= newData["System Memory GB"] + newData["System Memory MB"]/1000
newData.drop("System Memory MB", inplace = True, axis =1 )

...

sys_df = newData

q4 = sys_df




# Run but do not modify this code
print(sys_df.head())

# Run but do not modify this code
print(sys_df.tail())

   System Time second  System Memory GB System User ID System Machine ID
0                 NaN               NaN           yw22            Carrot
1                40.0             3.382              ?                 ?
2                 NaN             2.271            tp7         Asparagus
3                31.0             3.493              ?          Eggplant
4                37.0             3.411              ?         Asparagus
     System Time second  System Memory GB System User ID System Machine ID
995              1860.0             2.258            zm3         Asparagus
996              1860.0             3.403            bk4                 ?
997              2160.0             3.350           yw22            Carrot
998              2640.0             3.366           yw22         Asparagus
999              1200.0             3.490            bk4         Asparagus


<ipython-input-9-7c29d5a37544>:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  newData["System Time second"][400:] = newData["System Time second"][400:]*60
C:\Users\brady\anaconda3\lib\site-packages\pandas\core\indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


### Question 5 (12 points)
The `sys_df` dataframe from question 3 should now be a little easier to read and use. Answer the following questions about `sys_df`.

1. How many rows are missing data in the `System Machine ID` column?
2. What is the average value of `System Memory GB` among the rows that are missing data in the `System User ID` column? 
3. How many rows are missing data in both the `System Time second` and `System Memory GB` columns?

Note: It is not necessary to complete all of question 4 in order to answer some of these questions, and we will also look at your code for partial credit.

In [10]:
# Put your code to answer the question here, and store answers as the following variables
newData["System Machine ID"]= newData["System Machine ID"].replace("?", np.nan)
newData["System User ID"]= newData["System User ID"].replace("?", np.nan)

q5_1 = len(q4.index)-len(q4["System Machine ID"].dropna().index) # number of rows
q5_2 = newData.loc[newData["System User ID"].isnull()]["System Memory GB"].mean()# average System Memory GB

q5_3 = len(newData.loc[newData["System Time second"].isnull() & newData["System Memory GB"].isnull()])

q5_3

69

## Part 3: Wrangling FDA JSON Dataset 
In this part we work with a messy JSON dataset containing information about several drugs labels.

### Question 6 (12 points)
Below we import the `FDADrugLabel.json` file into the `labels` variable. This is the same dataset as the practice. The resulting Python object is somewhat messy; we encourage you to explore the data before answering the questions.

In [11]:
# Run but do not modify this code
import json
with open("FDADrugLabel.json", encoding ='utf-8') as f:
    labels = json.load(f)

labels[0]

{'package_label_principal_display_panel': ['label'],
 'active_ingredient': 'titanium dioxide',
 'warnings': ['1. Do not use in the following cases(Eczema and scalp wounds) 2.Side Effects 1)Due to the use of this druf if rash, irritation, itching and symptopms of hypersnesitivity occur dicontinue use and consult your phamacisr or doctor 3.General Precautions 1)If in contact with the eyes, wash out thoroughty with water If the symptoms are servere, seek medical advice immediately 2)This product is for exeternal use only. Do not use for internal use 4.Storage and handling precautions 1)If possible, avoid direct sunlight and store in cool and area of low humidity 2)In order to maintain the quality of the product and avoid misuse 3)Avoid placing the product near fire and store out in reach of children'],
 'inactive_ingredient': 'butylene glycol, water, etc',
 'drug_interactions': '',
 'effective_time': '20160321',
 'openfda': {'manufacturer_name': 'Nature and Health Beauty Co., Ltd.',
  'un

In [12]:
# Use this cell to explore (and you are welcome to add more cells as needed)


Answer the following questions.

1. Print the average number of key/value (or name/value) pairs for the drugs.
2. Print the list of all of the `manufacturer_names` without any other information. `manufacturer_names` are not a top level key/name, you will need to search for where they are located and how to extract them.
3. Print how many drugs contain the string `child` anywhere in their `warnings` (including as part of larger strings like `children`). `warnings` is a top level key/name, but some drugs are missing this data and the information is contained in strings within lists of length 1. 

In [19]:
# Put your code to answer the question here, and store answers as the following variables

#Checkkkkkk
pairs =0
for value in range(0,len(labels)):
    pairs += len(labels[value])

q6_1 = pairs/len(labels)# number of rows


totalmanufacturers = []
for value in range(0,len(labels)):
    totalmanufacturers.append(labels[value]["openfda"]["manufacturer_name"])
    
q6_2 = totalmanufacturers # average System Memory GB

child =0
for value in range(0,len(labels)):
    if "warnings" in labels[value]:
        for value2 in labels[value]["warnings"]:
            if "child" in value2:
                child+=1
        
        
        
q6_3 = child# number of missing rows

q6_1

19.25